DashWallet/Sources/UI/CrowdNode/Views/OutlinedTextField.swift

```diff
@@ -0,0 +1,102 @@
+    private var outerBorder: CAShapeLayer!
+    
+    var label: String = "" {
+        didSet {
+            labelControl.text = label
+        }
+    }
+    
+    var isError: Bool = false {
+        didSet {
+            if isError {
+                self.layer.borderColor = UIColor.systemRed.cgColor
+                outerBorder.fillColor = UIColor.systemRed.withAlphaComponent(0.1).cgColor
+                outerBorder.strokeColor = UIColor.systemRed.withAlphaComponent(0.2).cgColor
+            } else {
+                outerBorder.fillColor = UIColor.clear.cgColor
+                outerBorder.strokeColor = UIColor.dw_dashBlue().withAlphaComponent(0.2).cgColor
+                
+                if isEditing {
+                    self.layer.borderColor = UIColor.dw_dashBlue().cgColor
+                } else {
+                    self.layer.borderColor = borderColor.cgColor
+                }
+            }
+        }
+    }
+    
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+        
+        self.borderStyle = .none
+        self.layer.borderColor = UIColor.lightGray.cgColor
+        self.layer.borderWidth = 1
+        self.layer.cornerRadius = 12
+        
+        labelControl.textColor = .dw_secondaryText()
+        labelControl.font = .dw_regularFont(ofSize: 12)
+        labelControl.text = label
+        labelControl.translatesAutoresizingMaskIntoConstraints = false
+        labelControl.clipsToBounds = true
+        labelControl.frame = CGRect(x: 0, y: 0, width: self.frame.size.width, height: 18)
+        
+        self.addSubview(labelControl)
+              
+        labelControl.topAnchor.constraint(equalTo: self.topAnchor, constant: 10).isActive = true
+        labelControl.leftAnchor.constraint(equalTo: self.leftAnchor, constant: 15).isActive = true
+
+        let shape = CAShapeLayer()
+        shape.lineWidth = 4
+        shape.strokeColor = UIColor.dw_dashBlue().withAlphaComponent(0.2).cgColor
+        shape.fillColor = UIColor.clear.cgColor
+        outerBorder = shape
+        
+        self.addTarget(self, action: #selector(self.onEditingBegin), for: .editingDidBegin)
+        self.addTarget(self, action: #selector(self.onEditingEnd), for: .editingDidEnd)
+    }
+
+    override func textRect(forBounds bounds: CGRect) -> CGRect {
+        return bounds.inset(by: padding)
+    }
+
+    override func placeholderRect(forBounds bounds: CGRect) -> CGRect {
+        return bounds.inset(by: padding)
+    }
+
+    override func editingRect(forBounds bounds: CGRect) -> CGRect {
+        return bounds.inset(by: padding)
+    }
+    
+    @objc func onEditingBegin() {
+        let path = UIBezierPath(roundedRect: self.layer.bounds.inset(by: UIEdgeInsets(top: -2, left: -2, bottom: -2, right: -2)), cornerRadius: 14)
+        outerBorder.path = path.cgPath
+        self.layer.addSublayer(outerBorder)
+        self.layer.borderColor = UIColor.dw_dashBlue().cgColor
+    }
```

Comment:
I figured I'll keep the external border as a sublayer - that way we can implement the error state according to the design and have it all contained in a single control for reusability sake.
https://www.figma.com/file/azdJACb5WmivxYVhB5q46F/Design-system%E3%83%BBAndroid?node-id=3014%3A14441&t=K4J99XOQ6jVwlIWC-0

Reply:
@Syn-McJ Sound good — this approach works fine as well. The only thing is that we should not use `onEditingBegin` to calculate the rect for the path, this could be done within `layoutSubviews` or `drawRect` methods: 

```
override func layoutSubviews() {
    super.layoutSubviews()

    let path = UIBezierPath(roundedRect: bounds.inset(by: UIEdgeInsets(top: -2, left: -2, bottom: -2, right: -2)), cornerRadius: 14)
    outerBorder.path = path.cgPath
}
    
```

To change the visibility of `outerBorder` layer, I suggest just to toggle `isHidden` flag, when onEditingBegin or onEditingEnd is being invoked, instead of adding/removing for the view stack.

Reply:
Makes sense

Reply:
Fixed